### PR TITLE
Support code build path with dockerfile

### PR DESCRIPTION
**Pull request by RuoyuWang-MS**
What does this implement/fix? Explain your changes.
---------------------------------------------------
…
Add ACR
Use ACR task to build code with dockerfile

Does this close any currently open issues?
------------------------------------------

Any relevant logs, screenshots, error output, etc.?
-------------------------------------

Any other comments?
-------------------
…

Has this been tested?
---------------------------
- [ ] Tested

### CODE REVIEW

File: azure-toolkit-libs/azure-toolkit-containerregistry-lib/src/main/java/com/microsoft/azure/toolkit/lib/containerregistry/config/ContainerRegistryConfig.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+@Builder
+@Getter
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
```
Comment (allxiao) [inline azure-toolkit-libs/azure-toolkit-containerregistry-lib/src/main/java/com/microsoft/azure/toolkit/lib/containerregistry/config/ContainerRegistryConfig.java]: Check https://projectlombok.org/features/Data to make it a bit cleaner.

File: azure-toolkit-libs/azure-toolkit-containerregistry-lib/src/main/java/com/microsoft/azure/toolkit/lib/containerregistry/config/ContainerRegistryConfig.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
+package com.microsoft.azure.toolkit.lib.containerregistry.config;
+
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+@Builder
```
Comment (allxiao) [inline azure-toolkit-libs/azure-toolkit-containerregistry-lib/src/main/java/com/microsoft/azure/toolkit/lib/containerregistry/config/ContainerRegistryConfig.java]: Maybe we can remove Builder as we didn't call Builder methods anywhere.

File: azure-toolkit-libs/azure-toolkit-containerapps-lib/src/main/java/com/microsoft/azure/toolkit/lib/containerapps/task/DeployContainerAppTask.java [hunk 3]
```diff
@@ -78,6 +84,22 @@ private void addCreateAppEnvironmentTaskIfNecessary(@Nonnull final ContainerApps
         }
     }
 
+    //todo(ruoyuwang): check new ACR logic here. Reuse the existing ACR if it exists
+    private void addCreateContainerRegistryTaskIfNecessary(@Nonnull final ContainerRegistryConfig containerRegistryConfig) {
+        final ContainerRegistry registry = Azure.az(AzureContainerRegistry.class).registry(containerRegistryConfig.getSubscriptionId())
```
Comment (allxiao) [inline azure-toolkit-libs/azure-toolkit-containerapps-lib/src/main/java/com/microsoft/azure/toolkit/lib/containerapps/task/DeployContainerAppTask.java:89]: Maybe we can assume the customer will work in a single subscription in most of the cases. So we may have some kind of fallback logic for subscription configuration.

If the customer didn't provide the container registry subscription, we can default to the ACA subscription.

File: azure-container-apps-maven-plugin/src/main/java/com/microsoft/azure/maven/containerapps/parser/ConfigParser.java [hunk 4]
```diff
@@ -30,22 +39,31 @@ public ContainerAppConfig getContainerAppConfig() {
...
     //todo(ruoyuwang): whether we should support multi containers? CLI/Portal only support one container
     //todo(ruoyuwang): support container registry part
-    public ContainerAppDraft.ImageConfig getImageConfigFromContainers(List<AppContainerMavenConfig> containers) {
+    public ContainerAppDraft.ImageConfig getImageConfigFromContainers() {
+        List<AppContainerMavenConfig> containers = mojo.getContainers();
         if (containers == null || containers.isEmpty()) {
             return null;
         }
-        final ContainerAppDraft.ImageConfig imageConfig = new ContainerAppDraft.ImageConfig(containers.get(0).getImage());
+        final String defaultImageName = String.format("%s%s/%s:%s", mojo.getRegistry().getRegistryName(), ContainerRegistry.ACR_IMAGE_SUFFIX, mojo.getAppName(), timestamp);
```
Comment (allxiao) [inline azure-container-apps-maven-plugin/src/main/java/com/microsoft/azure/maven/containerapps/parser/ConfigParser.java:57]: To make it working for sovereign clouds, we need to read the suffix from somewhere.

Comment (allxiao) [reply azure-container-apps-maven-plugin/src/main/java/com/microsoft/azure/maven/containerapps/parser/ConfigParser.java:57]: Just curious, does IntelliJ plugin follows the same image tagging naming convention?

File: azure-container-apps-maven-plugin/src/main/java/com/microsoft/azure/maven/containerapps/parser/ConfigParser.java [hunk 5]
```diff
@@ -30,22 +39,31 @@ public ContainerAppConfig getContainerAppConfig() {
...
+        final String fullImageName = Optional.ofNullable(containers.get(0).getImage()).orElse(defaultImageName);
```
Comment (allxiao) [inline azure-container-apps-maven-plugin/src/main/java/com/microsoft/azure/maven/containerapps/parser/ConfigParser.java:58]: In CI scenario, it's not a good practice to reuse the same name. The container environment may set `imagePullPolicy=IfNotPresent`, which will prevent further update of the image with the same tag.

Better to check with tooling team on recommended experience here.

File: azure-toolkit-libs/azure-toolkit-containerapps-lib/src/main/java/com/microsoft/azure/toolkit/lib/containerapps/task/DeployContainerAppTask.java (hunk 3 above)
Comment (RuoyuWang-MS) [reply azure-toolkit-libs/azure-toolkit-containerapps-lib/src/main/java/com/microsoft/azure/toolkit/lib/containerapps/task/DeployContainerAppTask.java:89]: The registry default sub and rg will be set to the app sub and rg if not specified. [link](https://github.com/microsoft/azure-maven-plugins/pull/2495/files#:~:text=config.setSubscriptionId(Optional.ofNullable(mojo.getRegistry().getSubscriptionId()).orElse(mojo.getSubscriptionId()))%3B)

File: azure-container-apps-maven-plugin/src/main/java/com/microsoft/azure/maven/containerapps/parser/ConfigParser.java (hunk 4 above)
Comment (RuoyuWang-MS) [reply azure-container-apps-maven-plugin/src/main/java/com/microsoft/azure/maven/containerapps/parser/ConfigParser.java:57]: Will support sovereign clouds later. Now Intellij plugin also use this public cloud suffix and use appName:aca{timestamp} as the image tagging name. Here is an example: ruoyutest.azurecr.io/hello:aca-20241022135134356

File: azure-container-apps-maven-plugin/src/main/java/com/microsoft/azure/maven/containerapps/parser/ConfigParser.java (hunk 5 above)
Comment (RuoyuWang-MS) [reply azure-container-apps-maven-plugin/src/main/java/com/microsoft/azure/maven/containerapps/parser/ConfigParser.java:58]: Customers are likely to provide no image part in their config when using code/artifact type. We will generate the `defaultImageName` for them and won't reuse the name. The code just gives them an option to specify it. Most of the time, this scenario doesn't happen.

**Review by allxiao — APPROVED** [review]

**Pull request merged**